Adding a message to stop users from booking more days off than what t…

mcromie92 opened the pull request:
…hey have remaining
> calculated the approved days
>caculated the total days
> from this calculated the remaining holidays user has left 
> show message if the user tries to book more days than what they have allocated 
<img width="493" alt="capture1" src="https://user-images.githubusercontent.com/42771010/46413657-5b9e2200-c719-11e8-8a4b-ac230e2ca905.PNG">

mcromie92 committed:
adding a message to stop users from booking more days off than what they have remaining

girwin575 commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -52,6 +54,13 @@ const BookingModalForm = props => {
 
+  const remainingDays = (getTotalDays - getApprovedDays());
+  const composeRemainingHolsErrorMessage = () => {
+    if (remainingDays <= 0 || bookingDuration > remainingDays) { 
```
Is the situation of less than zero ever going to apply?

girwin575 commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -90,77 +99,83 @@ const BookingModalForm = props => {
+      {composeRemainingHolsErrorMessage() ?
+        <HolidayAlert>
+          <strong> {remainingDays} Days Remaining </strong>
+          <p> You do not have enough remaining holidays, please contact a member of the HR team. </p>     
```
If you could make this a tag rather than a strong tag then put `font-weight: bold;` in the styled component.

girwin575 commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -90,77 +99,83 @@ const BookingModalForm = props => {
-      </Form>
+      {composeRemainingHolsErrorMessage() ?
+        <HolidayAlert>
+          <strong> {remainingDays} Days Remaining </strong>
```
If you could make this a tag rather than a strong tag then put `font-weight: bold;` in the styled component.

girwin575 commented on web-app/src/components/BookingModal/BookingModalForm/styled.js:
```diff
@@ -18,3 +18,22 @@ export const NoticeAlert = styled.div`
 `;
+
+export const HolidayAlert = styled.div`
+margin: 15px 0;
```
The tabbing needs adjusted for the HolidayAlert block.

bgribz commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -52,6 +54,13 @@ const BookingModalForm = props => {
     }
   };
 
+  const remainingDays = (getTotalDays - getApprovedDays());
```
Should `getTotalDays` be called here rather than referenced? I.e. `getTotalDays() - getApprovedDays()`

bgribz commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -90,77 +99,83 @@ const BookingModalForm = props => {
+        <HolidayAlert>
+          <strong> {remainingDays} Days Remaining </strong>
+          <p> You do not have enough remaining holidays, please contact a member of the HR team. </p>     
+        </HolidayAlert> :
```
Having this much code displayed via a ternary is a good case for moving it to its own component

bgribz commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -52,6 +54,13 @@ const BookingModalForm = props => {
+  const remainingDays = (getTotalDays - getApprovedDays());
+  const composeRemainingHolsErrorMessage = () => {
+    if (remainingDays <= 0 || bookingDuration > remainingDays) { 
+      return true;
```
I think this could be named better as there's no composition going on in the function, and we can get rid of the `if` as well

```
const showRemainingHolsErrorMessage = () => {
  return (remainingDays <= 0 || bookingDuration > remainingDays)
}
```

d1wilko commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -90,77 +99,83 @@ const BookingModalForm = props => {
+              name: 'eventTypeId',
+              options: renderWFH(),
+            }}
+            
```
remove line break here?

William-Unosquare replied:
Can I take a moment @d1wilko to just appreciate this:
`('G' || 'B') + 'arry'`

mcromie92 committed:
fixing changes and adding in more functionality

mcromie92 commented:
@girwin575 @bgribz made changes requested here, added some more functionality too.
Previously if the user had 3 days remaining they could request two days twice and both could be approved.
I have added functionality to take into account how many days they have pending against the remaining days - so if they have three days remaining and 2 pending they will only be allowed to request one day. This means the remaining holidays will never go below 0.
Hope this makes sense, any questions let me know :)

d1wilko reviewed (commented):
some real good code, one tiny issue i noticed apart from what `('G' || 'B') + 'arry'` mentioned

William-Unosquare commented on web-app/src/components/BookingModal/BookingModalForm/container.js:
```diff
@@ -72,6 +75,22 @@ const Container = Wrapped =>
+      );
+    };
+    
+    getTotalDays = (this.props.userDetails.totalHolidays);
```
I don't think it's good practice to assign variables like this outside of function scope in a class. The naming of this also makes me think it's a function, especially alongside `getPendingDays` etc which are actually functions.

A far cleaner way to do this would be to destructure this value from props around line `183`.

William-Unosquare commented on web-app/src/components/BookingModal/BookingModalForm/container.js:
```diff
@@ -179,6 +200,7 @@ const Container = Wrapped =>
           }
           createEvent={event => createEvent(event, formData)}
           updateEvent={event => updateEvent(event, formData)}
+          getTotalDays={this.getTotalDays}
```
Rename this prop to "totalDays". Same reason as above.

William-Unosquare commented on web-app/src/components/BookingModal/BookingModalForm/container.js:
```diff
@@ -170,6 +189,8 @@ const Container = Wrapped =>
           capturedRejectionReponseText={capturedRejectionReponseText}
           assignRejectionResponseText={this.assignRejectionResponseText}
+          getApprovedDays={this.getApprovedDays()}
+          getPendingDays={this.getPendingDays()}
```
Rename these to "approvedDays" and "pendingDays". The function names are fine but the prop name makes these props sound like functions.

William-Unosquare commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -52,6 +55,22 @@ const BookingModalForm = props => {
   };
 
+  const remainingDays = (getTotalDays - getApprovedDays);
+  const pendingDays = getPendingDays;
```
Won't be any need for this line if you rename the props.

William-Unosquare commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -52,6 +55,22 @@ const BookingModalForm = props => {
+  const pendingDays = getPendingDays;
+  const availableDays = (remainingDays - pendingDays);
+  const showRemainingHolsAlertMessage = () => {
+    return (remainingDays === 0 || bookingDuration > remainingDays) || (availableDays < bookingDuration);     
```
Is the `remainingDays === 0` part needed? I think `bookingDuration > remainingDays` will cover this saying it's not possible to book 0 day holidays.

William-Unosquare commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -90,77 +109,79 @@ const BookingModalForm = props => {
-        />
-      </Form>
+      {showRemainingHolsAlertMessage() ? holidayAlertMessage() :
+        <Form formData={formData} formStatus={formStatus} actions={createCtas()}>
```
Pull the Form out into its own function.  A lot being rendered here off the ternary.

William-Unosquare commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -175,6 +196,9 @@ BookingModalForm.propTypes = {
   isSameDay: PT.bool.isRequired,
+  getApprovedDays: PT.number,
+  getPendingDays: PT.number,
+  getTotalDays: PT.number,
```
.isRequired

William-Unosquare reviewed (changes requested):
Few things to change here, mostly around naming of variables.

girwin575 commented on web-app/src/components/BookingModal/BookingModalForm/index.js:
```diff
@@ -52,6 +55,22 @@ const BookingModalForm = props => {
+  const holidayAlertMessage = () => {
+    return (
+      <HolidayAlert>
+        <tag>{availableDays} Days Available</tag>
```
I may be silly here but I've never seen the <tag> element before and it's not on https://www.w3schools.com/tags/ does it exist, if so what purpose does it serve?
